### ISSUE TITLE
name clashes not handled when using multiple root_parents directories [Tigris #302]

### ISSUE THREAD

**Issue by cmpilato** [viewvc/viewvc#58]
## Description

```
When you have multiple root parents, and the same repository name occurs in more
than one parent directory, only the last found repository will be accessible;
the others will be shadowed.

Furthermore, if the file hierarchy is deeper than 1 level, it is not sufficient
to use only the parent directory name to form a proper root name for mapping
back to a Subversion repository URL (base+root+repository).

For example, if you have a hypothetical (big) project:

  root_parents = /srv/svn/repos/doc : svn,
                 /srv/svn/repos/engine : svn,
                 /srv/svn/repos/gui : svn,
                 /srv/svn/repos/media/images : svn,
                 /srv/svn/repos/media/sound : svn,
                 /srv/svn/repos/media/video : svn

with several repositories under each root, it is first of all likely that a
common repository name, e.g. "builds", will occur under several roots.
(Currently, only one of these will be visible.) Second, taking only the parent
directory name, e.g. "images" above, is not enough to map to a Subversion
server URL that might need the entire string "media/images". There is not even
necessarily a direct mapping between the file system paths and the URLs that
the Subversion server uses, so it should be possible to specify the desired root
name explicitly.
```
## Metadata Imported from Tigris (Issue 302)
- **Creation Date**: 2007-06-14 08:08:56
- **Reporter**: richardc
- **Subcomponent**: core
- **Version**: 1.0.4
- **Milestone**: _unscheduled_
- **Keywords**: 
- **Cc**: 
### 2007-06-14 08:16:18 by richardc

```
Created an attachment (id=154)
suggested patch to viewvc.py
```
### 2007-06-14 08:31:39 by richardc

```
The attached patch solves these problems (at least, it works for us).

First, it automatically incorporates the parent directory name in the root name,
so that two repositories "/srv/svn/repos/foo/x" and "/srv/svn/repos/bar/x" will
both be available, and be listed as "foo/x" and "bar/x" by default. 

Second, it makes it possible to add an optional explicit prefix field for
each entry in root_parents, so that the following works:

  root_parents = /srv/svn/repos/doc : svn,
                 /srv/svn/repos/engine : svn,
                 /srv/svn/repos/gui : svn,
                 /srv/svn/repos/media/images : media/images : svn,
                 /srv/svn/repos/media/sound : media/sound : svn,
                 /srv/svn/repos/media/video : media/video : svn

where e.g. a repository "x" under "/srv/svn/repos/engine" will be shown as
"engine/x", while a repository "x" under "/srv/svn/repos/media/sound" will be
shown as "media/sound/x".

It is also be possible to specify an empty prefix ("path : : svn"), for those
who know that their repository names do not clash even if they are in separate
parent directories, and do not want the parent directory names to be shown.
```
### 2007-08-07 15:34:01 by jpcollins

```
doesn't there need to be a second part to this patch (in the Request class) to
handle these multi-leveled roots in the incoming requests?
```
### 2007-08-13 01:25:37 by richardc

```
As far as I can tell, the rest of the code does not care whether an entry in 
svn_roots is a single name "foo" or contains a slash as in "foo/bar". Like I
said, it works for us, apparently with no problems, and the patch above is the
only change.
```
### 2009-02-26 07:52:51 by cmpilato

```
Moving reviewed-but-not-scheduled issues into a new milestone (to distinguish
them from unreviewed one).
```
### viewvc.patch - suggested patch to viewvc.py

_Posted 2007-06-14 08:16:18 by richardc_

```
--- /usr/local/lib/viewvc-1.0.4/lib/viewvc-orig.py  2007-04-10 08:16:01.000000000 +0200
+++ /usr/local/lib/viewvc-1.0.4/lib/viewvc.py   2007-06-12 16:46:06.000000000 +0200
@@ -3607,6 +3610,13 @@
     repo_type = string.strip(pp[pos+1:])
     pp = os.path.normpath(string.strip(pp[:pos]))

+    pos = string.rfind(pp, ':')
+    if pos < 0:
+        base = os.path.basename(pp)
+    else:
+        base = string.strip(pp[pos+1:])
+        pp = os.path.normpath(string.strip(pp[:pos]))
+
     try:
       subpaths = os.listdir(pp)
     except OSError:
@@ -3622,9 +3632,9 @@
             and (os.path.exists(os.path.join(pp, subpath, "CVSROOT", "config"))
                  or (cvsroot and (subpath != 'CVSROOT'
                                   or not cfg.options.hide_cvsroot)))):
-          cfg.general.cvs_roots[subpath] = os.path.join(pp, subpath)
+          cfg.general.cvs_roots[base+"/"+subpath] = os.path.join(pp, subpath)
         elif repo_type == 'svn' and \
              os.path.exists(os.path.join(pp, subpath, "format")):
-          cfg.general.svn_roots[subpath] = os.path.join(pp, subpath)
+          cfg.general.svn_roots[base+"/"+subpath] = os.path.join(pp, subpath)

   debug.t_end('load-config')
```

**Comment by rkraats** [comment]
I created a patch against the current version (which also supports the 'root as url component' functionality) in #135

**Comment by cmpilato** [comment]
I'm considering this issue closed thanks to the changes merge via #135 .
